### PR TITLE
Issue #11132 - adding hookName to beforeHook and afterHook

### PR DESCRIPTION
**Pull request by HananArgov**
## Proposed changes

Adding hookName to `beforeHook` and `afterHook` per issue #11132 

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Further comments

as seen in the issue, this will able us to see which hook is referred in `beforeHook` and `afterHook`

### CODE REVIEW

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts [hunk 1]
```diff
@@ -34,7 +35,8 @@ export const testFnWrapper = function (
         BeforeHookParam<unknown>,
         AfterHookParam<unknown>,
         string,
-        number
+        number,
+        string?
```
Comment (erwinheitzman) [inline packages/wdio-utils/src/test-framework/testFnWrapper.ts:39]: From what I can tell, this isn't optional and is always passed

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts [hunk 2]
```diff
@@ -60,10 +63,14 @@ export const testFrameworkFnWrapper = async function (
     { beforeFn, beforeFnArgs }: BeforeHookParam<unknown>,
     { afterFn, afterFnArgs }: AfterHookParam<unknown>,
     cid: string,
-    repeatTest = 0
+    repeatTest = 0,
+    hookName?: string
 ) {
     const retries = { attempts: 0, limit: repeatTest }
     const beforeArgs = beforeFnArgs(this)
+    if (type !== 'Test') {
```
Comment (erwinheitzman) [inline packages/wdio-utils/src/test-framework/testFnWrapper.ts:71]: Let's explicitly add it to the functions we want instead of adding it to everything but Test

**Review by erwinheitzman — COMMENTED** [review]
Nice going 👍

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts (hunk 2 above)
Comment (HananArgov) [reply packages/wdio-utils/src/test-framework/testFnWrapper.ts:71]: i'm not sure if i understood what you meant, or know how implement.

do you mean that i'll need to remove the if, so it would work with `type==='Test'` as well? or that when we push to this to `afterArgs/beforeArgs` the hookName will be inserted as well, as part of line 70. 

if it's the second one, i'm a bit stumped here, since i can't quite understand how to do this (help wanted 🙂)

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts (hunk 1 above)
Comment (HananArgov) [reply packages/wdio-utils/src/test-framework/testFnWrapper.ts:39]: i'm leaving this optional, since `runSpec` in `testInterfaceWrapper` calls it as well, and then we'll add an unused argument (`origFn.name` is `it`)

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts (hunk 2 above)
Comment (HananArgov) [reply packages/wdio-utils/src/test-framework/testFnWrapper.ts]: changed the comparison to look for `Hook`, since we want to implement this on all hooks

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts [hunk 3]
```diff
@@ -81,7 +88,9 @@ export const testFrameworkFnWrapper = async function (
     }
     const duration = Date.now() - testStart
     const afterArgs = afterFnArgs(this)
-
+    if (type === 'Hook') {
+        beforeArgs.push(hookName)
```
Comment (erwinheitzman) [inline packages/wdio-utils/src/test-framework/testFnWrapper.ts:92]: I believe this should be
```suggestion
        afterArgs.push(hookName)
```

**Review by erwinheitzman — CHANGES_REQUESTED** [review]
Left another comment.

We also have to update the docs and templates with the new argument

**Comment by HananArgov** [comment]
https://github.com/webdriverio/webdriverio/pull/11148#pullrequestreview-1622252641
i've updated the templates and documents, except for the i18n, since I don't know the other languages

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts [hunk 4]
```diff
@@ -60,10 +63,14 @@ export const testFrameworkFnWrapper = async function (
     { beforeFn, beforeFnArgs }: BeforeHookParam<unknown>,
     { afterFn, afterFnArgs }: AfterHookParam<unknown>,
     cid: string,
-    repeatTest = 0
+    repeatTest = 0,
+    hookName?: string
 ) {
     const retries = { attempts: 0, limit: repeatTest }
     const beforeArgs = beforeFnArgs(this)
+    if (type === 'Hook') {
```
Comment (christian-bromann) [inline packages/wdio-utils/src/test-framework/testFnWrapper.ts:71]:
```suggestion
    if (type === 'Hook' && hookName) {
```

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts [hunk 5]
```diff
@@ -81,7 +88,9 @@ export const testFrameworkFnWrapper = async function (
     }
     const duration = Date.now() - testStart
     const afterArgs = afterFnArgs(this)
-
+    if (type === 'Hook') {
```
Comment (christian-bromann) [inline packages/wdio-utils/src/test-framework/testFnWrapper.ts:91]:
```suggestion
    if (type === 'Hook' && hookName) {
```

**Review by christian-bromann — COMMENTED** [review]
It would be nice to have a test for this to ensure we verify this functionality!

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts (hunk 5 above)
Comment (HananArgov) [reply packages/wdio-utils/src/test-framework/testFnWrapper.ts]: done

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts (hunk 4 above)
Comment (HananArgov) [reply packages/wdio-utils/src/test-framework/testFnWrapper.ts]: done

File: packages/wdio-types/src/Services.ts [hunk 6]
```diff
@@ -234,17 +234,19 @@ export interface HookFunctions {
...
      * Hook that gets executed _after_ a hook within the suite ends (e.g. runs after calling
      * afterEach in Mocha). `stepData` and `world` are Cucumber framework specific.
      * @param test      details to current running test (represents step in Cucumber)
      * @param context   context to current running test (represents World object in Cucumber)
+     * @param hookName  name of the hook
      * @param result    test result
-     */
-    afterHook?(test: Test, context: any, result: TestResult): unknown | Promise<unknown>
+    */
+    afterHook?(test: Test, context: any, hookName: string, result: TestResult): unknown | Promise<unknown>
```
Comment (christian-bromann) [inline packages/wdio-types/src/Services.ts:249]: This could be a potentially breaking change for user that access the `result` object, how about we also add it as last argument?

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts [hunk 7]
```diff
@@ -81,7 +88,9 @@ export const testFrameworkFnWrapper = async function (
     }
     const duration = Date.now() - testStart
     const afterArgs = afterFnArgs(this)
-
+    if (type === 'Hook' && hookName) {
+        afterArgs.push(hookName)
+    }
```
Comment (christian-bromann) [inline packages/wdio-utils/src/test-framework/testFnWrapper.ts:93]: which means pushing this one a bit down

**Review by christian-bromann — COMMENTED** [review]
One comment on the position of the argument in the `afterHook` interface. Also, I still would like to see a smoke test that verifies this feature exist.

File: packages/wdio-types/src/Services.ts (hunk 6 above)
Comment (HananArgov) [reply packages/wdio-types/src/Services.ts:249]: i'm trying to do so, but i'm getting an compiling issue with this part:
image

any idea on what to do here?

Comment (HananArgov) [reply packages/wdio-types/src/Services.ts:249]:
```
        const emitHookEvent = (
            fnName: string,
            eventType: string
        ) => (
            _test: never,
            _context: never,
            { error }: { error?: jasmine.FailedExpectation } = {},
            _hookName: never
```

Comment (christian-bromann) [reply packages/wdio-types/src/Services.ts:249]: Just leave out the parameter

Comment (erwinheitzman) [reply packages/wdio-types/src/Services.ts:249]: Small reminder to update the docs/templates since the argument is now last

**Comment by erwinheitzman** [comment]
Ping @HananArgov

**Comment by HananArgov** [comment]
sorry, i've been on vacation abroad for a while, and came back today.
i've done the changes, please review them again, hopefully I didn't miss anything

**Comment by christian-bromann** [comment]
@HananArgov it seems that the coverage dropped, feel free to update the [branch coverage](https://github.com/webdriverio/webdriverio/blob/main/vitest.config.ts#L34) to `88`%

File: packages/wdio-browserstack-service/src/service.ts [hunk 8]
```diff
@@ -176,7 +176,7 @@ export default class BrowserstackService implements Services.ServiceInstance {
         await this._insightsHandler?.beforeHook(test, context)
     }
 
-    async afterHook (test: Frameworks.Test|CucumberHook, context: unknown, result: Frameworks.TestResult) {
+    async afterHook (test: Frameworks.Test|CucumberHook, context: unknown, hookName: string,  result: Frameworks.TestResult) {
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/service.ts:179]: Shouldn't `hookName` come last?

File: examples/wdio/vite-vue-example/wdio.conf.ts [hunk 9]
```diff
@@ -242,13 +242,13 @@ export const config: Options.Testrunner = {
...
      */
-    // beforeHook: function (test, context) {
+    // beforeHook: function (test, context, hookName) {
     // },
     /**
      * Hook that gets executed _after_ a hook within the suite starts (e.g. runs after calling
      * afterEach in Mocha)
      */
-    // afterHook: function (test, context, { error, result, duration, passed, retries }) {
+    // afterHook: function (test, context, { error, result, duration, passed, retries }, hookName) {
```
Comment (christian-bromann) [inline examples/wdio/vite-vue-example/wdio.conf.ts:251]: according to these docs

File: website/i18n/bg/docusaurus-plugin-content-docs/current/Configuration.md [hunk 10]
```diff
@@ -66,7 +66,7 @@ Next to the WebDriver based capabilities you can apply browser and vendor specif
 
 Additionally, a useful utility is the Sauce Labs [Automated Test Configurator](https://wiki.saucelabs.com/display/DOCS/Platform+Configurator#/), which helps you create this object by clicking together your desired capabilities.
 
-Type: `Object`<br /> Default: `null`
```
Comment (christian-bromann) [inline website/i18n/bg/docusaurus-plugin-content-docs/current/Configuration.md:69]: Can we revert the changes in this file?

File: packages/wdio-browserstack-service/src/service.ts (hunk 8 above)
Comment (HananArgov) [reply packages/wdio-browserstack-service/src/service.ts:179]: i've tried but i'm getting the `unused var` lint error, which is weird, since when the arg is in this location, the lint doesn't show the error
image

Comment (christian-bromann) [reply packages/wdio-browserstack-service/src/service.ts:179]: The service doesn't need the parameter to be set, so you can just not add it there.

File: website/i18n/bg/docusaurus-plugin-content-docs/current/Configuration.md (hunk 10 above)
Comment (HananArgov) [reply website/i18n/bg/docusaurus-plugin-content-docs/current/Configuration.md]: u've reverted this. I don't know what changed this. maybe the RUNME extension?

File: packages/wdio-browserstack-service/src/service.ts (hunk 8 above)
Comment (HananArgov) [reply packages/wdio-browserstack-service/src/service.ts]: I removed the parameter

**Comment by christian-bromann** [comment]
> is there a chance maybe to arrange a video meeting to try and solve this?

Sure, please feel free to block a slot in the [Open Office Hours](https://webdriver.io/community/openofficehours).

**Review by christian-bromann — APPROVED** [review]
Nice work, LGTM 👍
